File: BatchTestTool/java/java automated tester/Test/compile_passed/solutions/HRManagementSystem.java
import java.util.HashMap;
import java.util.Map;

public class HRManagementSystem {
    private Map<Integer, Employee> employees;
    private HashMap<Integer, String> employeesMap;
    private Map<Integer, Map<String, String>> employeesInfo;

    public HRManagementSystem() {
        this.employees = new HashMap<>();
        this.employeesMap = new HashMap<>();
        this.employeesInfo = new HashMap<>();
    }

    public boolean addEmployee(int employeeId, String name, String position, String department, double salary) {
        if (employees.containsKey(employeeId)) {
            return false;
        } else {
            Employee employee = new Employee(name, position, department, salary);
            employees.put(employeeId, employee);
            return true;
        }
    }

    public boolean removeEmployee(int employeeId) {
        if (employeesMap.containsKey(employeeId)) {
            employeesMap.remove(employeeId);
            return true;
        } else {
            return false;
        }
    }

    public Map<Integer, Map<String, String>> listEmployees() {
        Map<Integer, Map<String, String>> employeeData = new HashMap<>();
        if (!employeesInfo.isEmpty()) {
            for (Map.Entry<Integer, Map<String, String>> entry : employeesInfo.entrySet()) {
                int employeeId = entry.getKey();
                Map<String, String> employeeInfo = entry.getValue();
                Map<String, String> employeeDetails = new HashMap<>();
                employeeDetails.put("employee_ID", String.valueOf(employeeId));
                for (Map.Entry<String, String> infoEntry : employeeInfo.entrySet()) {
                    employeeDetails.put(infoEntry.getKey(), infoEntry.getValue());
                }
                employeeData.put(employeeId, employeeDetails);
            }
        }
        return employeeData;
    }

    public boolean updateEmployee(int employeeId, Map<String, String> employeeInfo) {
        Map<String, String> employee = getEmployee(employeeId);
        if (employee == null) {
            return false;
        } else {
            for (String key : employeeInfo.keySet()) {
                if (!employee.containsKey(key)) {
                    return false;
                }
            }
            employee.putAll(employeeInfo);
            return true;
        }
    }

    public Map<String, String> getEmployee(int employeeId) {
        return employeesInfo.get(employeeId);
    }

    private static class Employee {
        private String name;
        private String position;
        private String department;
        private double salary;

        public Employee(String name, String position, String department, double salary) {
            this.name = name;
            this.position = position;
            this.department = department;
            this.salary = salary;
        }
    }
}

File: ClassEval_T/cpp/solution/HRManagementSystem.cpp
#include <string>
#include <map>
#include <variant>
#include <optional>
#include<unordered_map>

class HRManagementSystem {
public:
    HRManagementSystem();
    bool add_employee(int employee_id, const std::string& name, const std::string& position, const std::string& department, int salary);
    bool remove_employee(int employee_id);
    bool update_employee(int employee_id, const std::unordered_map<std::string, std::variant<std::string, int>>& employee_info);
    std::optional<std::unordered_map<std::string, std::variant<std::string, int>>> get_employee(int employee_id) const;
    std::unordered_map<int, std::unordered_map<std::string, std::variant<std::string, int>>> employees;
    std::unordered_map<int, std::unordered_map<std::string, std::variant<std::string, int>>> list_employees() const;
};

HRManagementSystem::HRManagementSystem() : employees() {}

bool HRManagementSystem::add_employee(int employee_id, const std::string& name, const std::string& position, const std::string& department, int salary) {
    if (employees.find(employee_id) != employees.end()) {
        return false;
    }
    else {
        employees[employee_id] = {
            {"name", name},
            {"position", position},
            {"department", department},
            {"salary", salary}
        };
        return true;
    }
}

bool HRManagementSystem::remove_employee(int employee_id) {
    auto it = employees.find(employee_id);
    if (it != employees.end()) {
        employees.erase(it);
        return true;
    }
    else {
        return false;
    }
}
bool HRManagementSystem::update_employee(int employee_id, const std::unordered_map<std::string, std::variant<std::string, int>>& employee_info) {
    auto employee_it = employees.find(employee_id);
    if (employee_it == employees.end()) {
        return false;
    }

    for (const auto& [key, value] : employee_info) {
        if (employee_it->second.find(key) == employee_it->second.end()) {
            return false;
        }
    }

    for (const auto& [key, value] : employee_info) {
        employee_it->second[key] = value;
    }

    return true;
}

std::optional<std::unordered_map<std::string, std::variant<std::string, int>>> HRManagementSystem::get_employee(int employee_id) const {
    auto it = employees.find(employee_id);
    if (it != employees.end()) {
        return it->second;
    }
    return std::nullopt;
}

std::unordered_map<int, std::unordered_map<std::string, std::variant<std::string, int>>> HRManagementSystem::list_employees() const {
    std::unordered_map<int, std::unordered_map<std::string, std::variant<std::string, int>>> employee_data;

    for (const auto& [employee_id, employee_info] : employees) {
        std::unordered_map<std::string, std::variant<std::string, int>> employee_details;
        employee_details["employee_ID"] = employee_id;
        for (const auto& [key, value] : employee_info) {
            employee_details[key] = value;
        }
        employee_data[employee_id] = employee_details;
    }

    return employee_data;
}

